File: slam/steps/ppkCorrection.mjs
// Порт PPK_CORRECTION из slamcloude/worker/pipeline/tasks.py (505-537) +
// gnss.py (parse_pos, rnx2rtkp_available, run_rnx2rtkp, fixed_ratio).
// Best-effort, как и в оригинале: рован/базовые обсервации редки для
// bag-пути S20 (там своя RTK-траектория из bag, см. decodeRaw.mjs) — этот
// шаг предназначен для будущих не-bag/внешних PPK-сценариев, отсутствие
// нужных входов просто пропускает шаг, не валит пайплайн.
import { execFileSync } from "node:child_process";
import { existsSync, mkdirSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { runStep } from "../lib/stepContext.mjs";

const Q_FIXED = 1;

function rnx2rtkpAvailable() {
  try {
    execFileSync("where", ["rnx2rtkp.exe"], { stdio: "pipe" });
    return true;
  } catch {
    return false;
  }
}

/** Parses an RTKLIB-style .pos file — direct port of gnss.parse_pos. */
export function parsePos(path) {
  const text = readFileSync(path, "utf8");
  const times = [], lats = [], lons = [], heights = [], quality = [];
  for (const rawLine of text.split("\n")) {
    const line = rawLine.trim();
    if (!line || line.startsWith("%") || line.startsWith("#")) continue;
    const parts = line.split(/\s+/);
    if (parts.length < 6) continue;
    const [datePart, timePart, lat, lon, h, q] = parts;
    const iso = `${datePart.replace(/\//g, "-")}T${timePart}Z`;
    const t = Date.parse(iso) / 1000;
    if (Number.isNaN(t)) continue;
    times.push(t); lats.push(parseFloat(lat)); lons.push(parseFloat(lon));
    heights.push(parseFloat(h)); quality.push(parseInt(q, 10));
  }
  const order = times.map((_, i) => i).sort((a, b) => times[a] - times[b]);
  const pick = (arr) => order.map((i) => arr[i]);
  return { times: pick(times), lats: pick(lats), lons: pick(lons), heights: pick(heights), quality: pick(quality) };
}

export function fixedRatio(traj) {
  if (!traj.quality.length) return 0;
  return traj.quality.filter((q) => q === Q_FIXED).length / traj.quality.length;
}

await runStep(async (ctx) => {
  const inputs = new Map((ctx.inputs || []).map((i) => [i.kind, i.path]));
  const hasAll = inputs.has("base_rinex") && inputs.has("rover_obs") && inputs.has("trajectory");
  if (!hasAll || !rnx2rtkpAvailable()) {
    return {}; // best-effort skip — matches tasks.py's early StepOutcome()
  }

  const outPos = join(ctx.scan_dir, "intermediate", "corrected.pos");
  mkdirSync(dirname(outPos), { recursive: true });
  const args = ["-p", "2", "-o", outPos, inputs.get("rover_obs"), inputs.get("base_rinex")];
  if (inputs.has("nav")) args.push(inputs.get("nav"));
  execFileSync("rnx2rtkp.exe", args, { stdio: "pipe", timeout: 3600_000 });

  if (!existsSync(outPos)) throw new Error("rnx2rtkp reported success but output .pos is missing");
  const corrected = parsePos(outPos);
  return { rtk_fixed: fixedRatio(corrected) >= 0.5 };
});

File: pyserver/apps/slam/pipeline/steps/ppk_correction.py
"""
Port of PPK_CORRECTION (tasks.py:505-537 + gnss.py's parse_pos,
rnx2rtkp_available, run_rnx2rtkp, fixed_ratio). Best-effort, as in the
original: rover/base observations are rare for the S20 bag path (which has
its own RTK trajectory extracted from the bag, see decode_raw.py) -- this
step is meant for future non-bag/external-PPK scenarios; missing required
inputs just skips the step rather than failing the pipeline.
"""

import shutil
import subprocess
from datetime import datetime, timezone
from pathlib import Path

from django.conf import settings

Q_FIXED = 1


def rnx2rtkp_available() -> bool:
    return shutil.which(settings.RNX2RTKP_EXE) is not None


def parse_pos(path) -> dict:
    """Parses an RTKLIB-style .pos file -- direct port of gnss.parse_pos."""
    rows = []
    with open(path, "r", encoding="utf-8") as f:
        for raw_line in f:
            line = raw_line.strip()
            if not line or line.startswith("%") or line.startswith("#"):
                continue
            parts = line.split()
            if len(parts) < 6:
                continue
            date_part, time_part, lat, lon, h, q = parts[:6]
            try:
                dt = datetime.strptime(f"{date_part} {time_part}", "%Y/%m/%d %H:%M:%S.%f").replace(tzinfo=timezone.utc)
            except ValueError:
                continue
            rows.append((dt.timestamp(), float(lat), float(lon), float(h), int(q)))
    rows.sort(key=lambda r: r[0])
    return {
        "times": [r[0] for r in rows], "lats": [r[1] for r in rows], "lons": [r[2] for r in rows],
        "heights": [r[3] for r in rows], "quality": [r[4] for r in rows],
    }


def fixed_ratio(traj: dict) -> float:
    if not traj["quality"]:
        return 0.0
    return sum(1 for q in traj["quality"] if q == Q_FIXED) / len(traj["quality"])


def run(ctx: dict) -> dict:
    inputs = {i["kind"]: i["path"] for i in (ctx.get("inputs") or [])}
    has_all = "base_rinex" in inputs and "rover_obs" in inputs and "trajectory" in inputs
    if not has_all or not rnx2rtkp_available():
        return {}  # best-effort skip -- matches tasks.py's early StepOutcome()

    out_pos = Path(ctx["scan_dir"]) / "intermediate" / "corrected.pos"
    out_pos.parent.mkdir(parents=True, exist_ok=True)
    args = [settings.RNX2RTKP_EXE, "-p", "2", "-o", str(out_pos), inputs["rover_obs"], inputs["base_rinex"]]
    if "nav" in inputs:
        args.append(inputs["nav"])
    subprocess.run(args, capture_output=True, timeout=3600, check=True)

    if not out_pos.is_file():
        raise RuntimeError("rnx2rtkp reported success but output .pos is missing")
    corrected = parse_pos(out_pos)
    return {"rtk_fixed": fixed_ratio(corrected) >= 0.5}

